File: zziri/ps/leetcode/Problems/3. Longest Substring Without Repeating Characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        l = len(s)
        if l == 1:
            return 1
        elif l == 0:
            return 0
        max_len = 0
        for i in range(l-1):
            d = dict()
            for j in range(i, l):
                if d.get(s[j]):
                    max_len = max(max_len, j-i)
                    break
                else:
                    d[s[j]] = True
                if j == l-1:
                    max_len = max(max_len, j-i+1)
                    break
        return max_len

File: zziri/ps/leetcode/problems/v1/3. Longest Substring Without Repeating Characters.java
// MEMO : 슬라이딩 윈도우 연습하자. 어렵다..

// Runtime: 6 ms, faster than 62.62% of Java online submissions for Longest Substring Without Repeating Characters.
// Memory Usage: 39.6 MB, less than 27.71% of Java online submissions for Longest Substring Without Repeating Characters.
class Solution {
    public int lengthOfLongestSubstring(String s) {
        Set<Character> set = new HashSet<>();
        int max = 0;
        int i = 0;
        for (int j=0; j<s.length(); j++) {
            char ch = s.charAt(j);
            
            while (set.contains(ch)) {
                set.remove(s.charAt(i++));
            }
            
            set.add(ch);
            max = Math.max(set.size(), max);
        }
        return max;
    }
}
